src/lib/optimizer/strategy/stored_table_column_alignment_rule.cpp

```diff
@@ -0,0 +1,63 @@
+#include "optimizer/strategy/stored_table_column_alignment_rule.hpp"
+
+#include <algorithm>
+#include <memory>
+
+#include "logical_query_plan/abstract_lqp_node.hpp"
+
+namespace opossum {
+
+void StoredTableColumnAlignmentRule::apply_to(const std::shared_ptr<AbstractLQPNode>& root) const {
+  StoredTableNodeUnorderedMap<std::vector<std::shared_ptr<StoredTableNode>>> gathered_stored_table_nodes_for_node;
+  StoredTableNodeUnorderedMap<std::vector<ColumnID>> aligned_pruned_column_ids_for_node;
+  recursively_gather_stored_table_nodes(root, gathered_stored_table_nodes_for_node, aligned_pruned_column_ids_for_node);
```

Comment:
The semantics of the map are a bit unclear to me. What is the key, what is the value?

Have you considered this approach:
1. Iterate over the LQP, store all StoredTableNodes in a column-pruning-agnostic unordered_set
2. For each set:
2.1 Iterate over the nodes and calculate the set intersection of the pruned column_ids.
2.2 Iterate over the nodes and set the new pruning list

Reply:
I thought about a similar approach where I use vectors instead of unordered sets. But sure, using sets makes more sense. I have abandoned the idea since I wanted to avoid the two iterations over the nodes for each vector/set. However, I think the code would be easier to understand using your mentioned approach. The computation costs for the two iterations should be negligible.

>What is the key, what is the value?

All StoredTableNodes with the same table name and pruned chunks are stored in a vector. `ColumnPruningAgnosticStoredTableNodeMap<std::vector<std::shared_ptr<StoredTableNode>>>
      gathered_stored_table_nodes_for_node;` is used to achieve that. To avoid your first mentioned iteration over all vectors/sets (2.1), the "final pruning list" for each visited StoredTableNode group is updated using `ColumnPruningAgnosticStoredTableNodeMap<std::vector<ColumnID>> aligned_pruned_column_ids_for_node`. As a final step, your mentioned 2.2-iteration is executed.

>Iterate over the LQP, store all StoredTableNodes in a column-pruning-agnostic unordered_set

But we somehow have to group the StoredTableNodes, i.e., put each StoredTableNode in the correct set. We actually don't need the current specialized hash generation and equals comparison. It would be sufficient to have a map with {table_name, pruned_chunks} as key and set<StoredTableNodes> as value. I'll try something like that soon.

Sorry for the long message.

Reply:
Ok, so instead of doing two passes where the first one results in the intersection, you perform the intersection on-the-fly. I'm not sure if this is more efficient. As you mentioned, the iteration itself shouldn't be a cost factor and the actual work is more or less the same, independent of where it happens.

> But we somehow have to group the StoredTableNodes, i.e., put each StoredTableNode in the correct set. We actually don't need the current specialized hash generation and equals comparison. It would be sufficient to have a map with {table_name, pruned_chunks} as key and sets of StoredTableNodes as values. I'll try something like that soon.

The specialized hash/equals is a good approach. My suggestion is simply to use a set with these specializations instead of a map. That way, we have a bucket of all StoredTableNodes that should be combined.